### PR TITLE
Fix fuzzer following gain map API changes

### PR DESCRIPTION
**Pull request by maryla-uc**
And a little bit of gain map reading cleanup in read.c (assume enableParsingGainMapMetadata is true in avifDecoderFindGainMapItem)

### CODE REVIEW

File: tests/gtest/avif_fuzztest_enc_dec_experimental.cc [hunk 1]
```diff
@@ -63,14 +63,12 @@ void EncodeDecodeValid(ImagePtr image, EncoderPtr encoder, DecoderPtr decoder) {
...
-  ASSERT_EQ(decoded_image->gainMap != nullptr,
-            decoder->gainMapPresent && (decoder->enableDecodingGainMap ||
-                                        decoder->enableParsingGainMapMetadata));
-  ASSERT_EQ(decoded_image->gainMap != nullptr &&
-                decoded_image->gainMap->image != nullptr,
-            decoder->gainMapPresent && decoder->enableDecodingGainMap);
+  if (decoder->enableParsingGainMapMetadata) {
+    EXPECT_EQ(decoder->gainMapPresent, image->gainMap != nullptr);
+  } else {
+    EXPECT_EQ(decoder->gainMapPresent, false);
```
Comment (wantehchang) [inline tests/gtest/avif_fuzztest_enc_dec_experimental.cc:69]: Nit: Use `EXPECT_FALSE`?

```
    EXPECT_FALSE(decoder->gainMapPresent);
```

**Review by wantehchang — APPROVED** [review]
Maryla: Please wait for Yannis's review because he reviewed your recent gain map pull requests.

File: src/read.c [hunk 2]
```diff
@@ -5422,18 +5417,13 @@ avifResult avifDecoderReset(avifDecoder * decoder)
...
                 const avifResult tmapParsingRes =
                     avifParseToneMappedImageBox(&decoder->image->gainMap->metadata, tmapData.data, tmapData.size, data->diag);
-                if (tmapParsingRes != AVIF_RESULT_OK) {
-                    avifGainMapDestroy(decoder->image->gainMap);
-                    decoder->image->gainMap = NULL;
-                }
                 if (tmapParsingRes == AVIF_RESULT_NOT_IMPLEMENTED) {
                     // Unsupported gain map version. Simply ignore the gain map.
+                    avifGainMapDestroy(decoder->image->gainMap);
+                    decoder->image->gainMap = NULL;
```
Comment (wantehchang) [inline src/read.c:5426]: If we return from the `AVIF_CHECKRES` at line 5428, is `decoder->image->gainMap` guaranteed to be NULL?

Comment (maryla-uc) [reply src/read.c:5426]: If AVIF_CHECKRES returns an error, decoder->image->gainMap might be set even if gainMapPresent is false (inconsistent state). But If the function returns an error, it doesn't have to guarantee anything about the output, does it?

File: tests/gtest/avif_fuzztest_helpers.h [hunk 3]
```diff
@@ -226,10 +226,11 @@ inline auto ArbitraryBaseAvifDecoder() {
 // options fuzzed, with the exception of 'ignoreColorAndAlpha' (because it would
 // break most tests' assumptions).
 inline auto ArbitraryAvifDecoderWithGainMapOptions() {
-  return fuzztest::Map(
-      AddGainMapOptionsToDecoder, ArbitraryBaseAvifDecoder(),
-      /*enable_parsing_gain_map_metadata=*/fuzztest::Arbitrary<bool>(),
-      /*enable_decoding_gain_map=*/fuzztest::Arbitrary<bool>());
+  return fuzztest::Map(AddGainMapOptionsToDecoder, ArbitraryBaseAvifDecoder(),
+                       /*gain_map_decode_mode=*/
+                       fuzztest::ElementOf({GainMapDecodeMode::kDontDecode,
```
Comment (y-guyon) [inline tests/gtest/avif_fuzztest_helpers.h:231]: Nit/optional: `fuzztest::Values` should be enough here

File: tests/gtest/avif_fuzztest_helpers.h [hunk 4]
```diff
@@ -226,10 +226,11 @@ inline auto ArbitraryBaseAvifDecoder() {
 // options fuzzed, with the exception of 'ignoreColorAndAlpha' (because it would
 // break most tests' assumptions).
 inline auto ArbitraryAvifDecoderWithGainMapOptions() {
-  return fuzztest::Map(
-      AddGainMapOptionsToDecoder, ArbitraryBaseAvifDecoder(),
-      /*enable_parsing_gain_map_metadata=*/fuzztest::Arbitrary<bool>(),
-      /*enable_decoding_gain_map=*/fuzztest::Arbitrary<bool>());
+  return fuzztest::Map(AddGainMapOptionsToDecoder, ArbitraryBaseAvifDecoder(),
+                       /*gain_map_decode_mode=*/
```
Comment (y-guyon) [inline tests/gtest/avif_fuzztest_helpers.h:230]: Nit/optional: I tend not to add the name of the field when the values are explicit enough

**Review by y-guyon — APPROVED** [review]

File: tests/gtest/avif_fuzztest_helpers.h (hunk 3 above)
Comment (maryla-uc) [reply tests/gtest/avif_fuzztest_helpers.h:230]: As discussed offline, there's no such function in fuzztest :)

**Pull request merged**